chain/mempool.go

```diff
@@ -12,6 +12,71 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
+// cachedInputs caches the inputs of the transactions in the mempool. This is
+// used to provide fast lookup between txids and inputs.
+type cachedInputs struct {
+	// inputs provides a fast lookup from input -> txid.
+	inputs map[wire.OutPoint]chainhash.Hash
+
+	// txids provides a fast lookup from txid -> inputs.
+	txids map[chainhash.Hash]map[wire.OutPoint]struct{}
+}
+
+// newCachedInputs creates a new cachedInputs.
+func newCachedInputs() *cachedInputs {
+	return &cachedInputs{
+		inputs: make(map[wire.OutPoint]chainhash.Hash),
+		txids:  make(map[chainhash.Hash]map[wire.OutPoint]struct{}),
+	}
+}
+
+// hasInput returns the txid and a boolean to indicate the given input is
+// found.
+func (c *cachedInputs) hasInput(op wire.OutPoint) (chainhash.Hash, bool) {
+	txid, ok := c.inputs[op]
+	return txid, ok
+}
+
+// addInput adds the given input to our cached inputs. If the input already
+// exists, the `inputs` map will be overwritten and the `txids` map will be
+// updated.
+func (c *cachedInputs) addInput(op wire.OutPoint, txid chainhash.Hash) {
+	oldTxid, ok := c.inputs[op]
```

Comment:
nit: `ok` not used. Or if it's used on L58, then it's not very clear whether it's overwritten in L50, so should maybe use a different name?